File: src.JAVA/src/swa/ariadne/outlines/bitmaps/BitmapOutlineShape.java
package swa.ariadne.outlines.bitmaps;

import java.awt.Dimension;
import java.awt.Point;
import java.awt.image.BufferedImage;
import java.util.Random;

import swa.ariadne.outlines.OutlineShape;
import swa.ariadne.outlines.OutlineShapeParameters;
import swa.util.Bitmap;
import swa.util.ImageResourceReader;
import swa.util.ImageUtil;

/**
 * An {@link OutlineShape} defined by the contour of a black and white bitmap image.
 */
public
class BitmapOutlineShape
extends OutlineShape
{
    //--------------------- Member variables and Properties

    /** A {@link Bitmap} with an underlying black-and-white image. */
    Bitmap bitmap;

    /** The location within this shape where the {@link #bitmap} is placed. */
    private final Point offset = new Point();

    /** False for a black-on-white image, true for a white-on-black image. */
    private final boolean hasBlackBackground;

    //--------------------- OutlineShape implementation

    @Override
    public boolean get(int x, int y)
    {
        int xi = x - offset.x, yi = y - offset.x;

        if (xi < 0 || yi < 0 || xi >= bitmap.getWidth() || yi >= bitmap.getHeight())
        {
            return hasBlackBackground;
        }
        else
        {
            return (bitmap.getBrightness(xi, yi) <= 0.5);
        }
    }

    //--------------------- Constructors

    /**
     * Constructor used by the {@link swa.ariadne.outlines.OutlineShapeFactory OutlineShapeFactory}.
     * Chooses one of the {@linkplain #chooseBitmap(Random) available bitmaps}}.
     * @param r A source of random numbers.
     * @param size Nominal size of the shape.
     * @param params Characteristic parameters of the shape: location and size.
     */
    public BitmapOutlineShape(Random r, Dimension size, OutlineShapeParameters params)
    {
        this(chooseBitmap(r), size, params);
    }

    /**
     * @param img A black and white image.
     * @param size Nominal size of the shape.
     * @param params Characteristic parameters of the shape: location and size.
     */
    private BitmapOutlineShape(BufferedImage img, Dimension size, OutlineShapeParameters params)
    {
        super(size);
        params.convertToShapeCoordinates(size);
        double scale = 2 * params.getSize() / Math.max(img.getWidth(), img.getHeight());

        BufferedImage scaledImg = ImageUtil.resize(ImageUtil.toBufferedImage(img), (int)(img.getWidth() * scale), (int)(img.getHeight() * scale));
        this.bitmap = new Bitmap(scaledImg);
        this.offset.x = (int)(params.getCenter().x - this.bitmap.getWidth() / 2.0);
        this.offset.y = (int)(params.getCenter().y - this.bitmap.getHeight() / 2.0);

        // [start] Determine the background color: black or white.

        int w = this.bitmap.getWidth() - 1, h = this.bitmap.getHeight() - 1;
        float cornerBrightness = 0;
        cornerBrightness += this.bitmap.getBrightness(0, 0);
        cornerBrightness += this.bitmap.getBrightness(w, 0);
        cornerBrightness += this.bitmap.getBrightness(0, h);
        cornerBrightness += this.bitmap.getBrightness(w, h);

        this.hasBlackBackground = (cornerBrightness < 0.5F * 4);

        // [end]
    }

    //--------------------- Static methods for creating a bitmap image

    /** The {@link ImageResourceReader} for this package. */
    private static final ImageResourceReader resourceReader = new ImageResourceReader(BitmapOutlineShape.class, "resources.txt");

    /**
     * @param r A source of random numbers.
     * @return One of the available bitmaps.
     */
    private static BufferedImage chooseBitmap(Random r)
    {
        return resourceReader.pickOne(r);
    }
}

File: src.CS/SWA.Ariadne.Outlines/BitmapOutlineShape.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing;

namespace SWA.Ariadne.Outlines
{
    /// <summary>
    /// An OutlineShape defined by the contour of a black and white bitmap image.
    /// </summary>
    internal class BitmapOutlineShape : OutlineShape
    {
        #region Class variables

        /// <summary>
        /// List of Property Methods that return bitmap images from the Resources file.
        /// </summary>
        private static List<System.Reflection.MethodInfo> BitmapProperties = SWA.Utilities.Resources.BitmapProperties(typeof(Resources.Bitmaps));

        #endregion

        #region Member variables and Properties

        private Bitmap map;
        private int mapXOffset, mapYOffset;
        private bool hasBlackBackground = false;

        /// <summary>
        /// Returns true if the image's pixel value is darker than 50%.
        /// Images are assumed to be painted black on a white background.
        /// </summary>
        /// <param name="x"></param>
        /// <param name="y"></param>
        /// <returns></returns>
        public override bool this[int x, int y]
        {
            get
            {
                int xi = x - mapXOffset, yi = y - mapYOffset;

                if (xi < 0 || yi < 0 || xi >= map.Width || yi >= map.Height)
                {
                    return hasBlackBackground;
                }
                else
                {
                    return (map.GetPixel(xi, yi).GetBrightness() <= 0.5);
                }
            }
        }

        #endregion

        #region Constructor

        /// <summary>
        /// Create an OutlineShape based on a Bitmap image.
        /// </summary>
        /// <param name="img"></param>
        /// <param name="xSize">width of the created shape</param>
        /// <param name="ySize">height of the created shape</param>
        /// <param name="centerX">X coordinate, relative to total width; 0.0 = top, 1.0 = bottom</param>
        /// <param name="centerY">Y coordinate, relative to total height; 0.0 = left, 1.0 = right</param>
        /// <param name="shapeSize">size, relative to distance of center from the border; 1.0 will touch the border</param>
        private BitmapOutlineShape(Bitmap img, int xSize, int ySize, double centerX, double centerY, double shapeSize)
            : base(xSize, ySize)
        {
            double xc, yc, sz;
            ConvertParameters(xSize, ySize, centerX, centerY, shapeSize, out xc, out yc, out sz);
            double scale = 2 * sz / Math.Max(img.Width, img.Height);

            this.map = new Bitmap(img, new Size((int)(img.Width * scale), (int)(img.Height * scale)));
            this.mapXOffset = (int)(xc - this.map.Width / 2.0);
            this.mapYOffset = (int)(yc - this.map.Height / 2.0);

            #region Determine the background color: black or white.

            int w = this.map.Width - 1, h = this.map.Height - 1;
            float cornerBrightness = 0;
            cornerBrightness += this.map.GetPixel(0, 0).GetBrightness();
            cornerBrightness += this.map.GetPixel(w, 0).GetBrightness();
            cornerBrightness += this.map.GetPixel(0, h).GetBrightness();
            cornerBrightness += this.map.GetPixel(w, h).GetBrightness();
            this.hasBlackBackground = (cornerBrightness < 0.5F * 4);

            #endregion
        }

        #endregion

        #region Static methods for creating OutlineShapes

        public static OutlineShape Random(Random r, int xSize, int ySize, double centerX, double centerY, double shapeSize)
        {
            System.Reflection.MethodInfo method = BitmapProperties[r.Next(BitmapProperties.Count)];
            Bitmap img = (Bitmap) method.Invoke(null, null);
            return new BitmapOutlineShape(img, xSize, ySize, centerX, centerY, shapeSize);
        }

        #endregion
    }
}
